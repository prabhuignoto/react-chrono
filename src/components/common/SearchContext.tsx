import React, { createContext, useState, useContext } from 'react';

interface SearchContextType {
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

// Create a context with default values
const SearchContext = createContext<SearchContextType>({
  searchTerm: '',
  setSearchTerm: () => {},
});

/**
 * Provider component for the search context
 */
export const SearchProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string>('');

  return (
    <SearchContext.Provider value={{ searchTerm, setSearchTerm }}>
      {children}
    </SearchContext.Provider>
  );
};

/**
 * Custom hook to use the search context
 */
export const useSearch = (): SearchContextType => {
  const context = useContext(SearchContext);

  if (context === undefined) {
    throw new Error('useSearch must be used within a SearchProvider');
  }

  return context;
};

export default SearchContext;
