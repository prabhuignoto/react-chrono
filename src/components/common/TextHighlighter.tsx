import React from 'react';
import styled from 'styled-components';
import { Theme } from '@models/Theme';

interface HighlightedTextProps {
  text: string;
  searchTerm: string;
  theme: Theme;
}

// Styled component for highlighted text
const HighlightSpan = styled.span<{ theme: Theme }>`
  background-color: ${({ theme }) =>
    theme.primary}33; /* Primary color with 20% opacity */
  color: ${({ theme }) => theme.primary};
  font-weight: 500;
  border-radius: 2px;
  padding: 0 2px;
`;

/**
 * Component that highlights occurrences of a search term within text
 */
const TextHighlighter: React.FC<HighlightedTextProps> = ({
  text,
  searchTerm,
  theme,
}) => {
  // If no search term or no text, just return the original text
  if (!searchTerm || !text) {
    return <>{text}</>;
  }

  // Escape special regex characters in the search term
  const escapedSearchTerm = searchTerm.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Create a case-insensitive regular expression to find all occurrences
  const regex = new RegExp(`(${escapedSearchTerm})`, 'gi');

  // Split the text by the search term
  const parts = text.split(regex);

  // Return the text with highlighted parts
  return (
    <>
      {parts.map((part, i) =>
        regex.test(part) ? (
          <HighlightSpan key={i} theme={theme}>
            {part}
          </HighlightSpan>
        ) : (
          <React.Fragment key={i}>{part}</React.Fragment>
        ),
      )}
    </>
  );
};

export default TextHighlighter;
