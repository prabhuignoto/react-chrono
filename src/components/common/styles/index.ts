import { Theme } from '@models/Theme';
import { css } from 'styled-components';

export const ScrollBar = css<{ theme: Theme }>`
  scrollbar-color: ${(p) => p.theme?.primary} default;
  scrollbar-width: thin;

  &::-webkit-scrollbar {
    width: 0.3em;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.2);
  }

  &::-webkit-scrollbar-thumb {
    background-color: ${(p) => p.theme?.primary};
    outline: 1px solid ${(p) => p.theme?.primary};
  }
`;
