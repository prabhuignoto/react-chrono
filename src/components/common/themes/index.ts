import { Theme } from '@models/Theme';

export const defaultTheme: Theme = {
  cardBgColor: '#ffffff',
  cardDetailsBackGround: '#ffffff',
  cardDetailsColor: '#000',
  cardMediaBgColor: '#f5f5f5',
  cardSubtitleColor: '#000',
  cardTitleColor: '#007FFF',
  detailsColor: '#000',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#f5f5f5',
  nestedCardDetailsBackGround: '#f5f5f5',
  nestedCardDetailsColor: '#000',
  nestedCardSubtitleColor: '#000',
  nestedCardTitleColor: '#000',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#f5f5f5',
  toolbarBtnBgColor: '#fff',
  toolbarTextColor: '#000',
};

export const darkTheme: Theme = {
  cardBgColor: '#191919',
  cardDetailsBackGround: '#191919',
  cardDetailsColor: '#ffff0f',
  cardMediaBgColor: '#2f2f2f',
  cardSubtitleColor: '#ffffff',
  cardTitleColor: '#007FFF',
  detailsColor: '#ffffff',
  iconBackgroundColor: '#007FFF',
  nestedCardBgColor: '#333333',
  nestedCardDetailsBackGround: '#333333',
  nestedCardDetailsColor: '#ffffff',
  nestedCardSubtitleColor: '#ffffff',
  nestedCardTitleColor: '#ffffff',
  primary: '#007FFF',
  secondary: '#ffdf00',
  titleColor: '#007FFF',
  titleColorActive: '#007FFF',
  toolbarBgColor: '#555',
  toolbarBtnBgColor: '#222',
  toolbarTextColor: '#ffffff',
};
