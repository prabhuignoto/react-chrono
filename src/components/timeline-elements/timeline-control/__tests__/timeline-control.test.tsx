import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';
import { commonProps, customRender, providerProps } from '../../../common/test';
import Controls from '../timeline-control';

describe('TimelineControl', () => {
  it('should render', () => {
    const { container } = customRender(<Controls {...commonProps} />, {
      providerProps,
    });
    expect(container).toMatchSnapshot();
  });

  // check if all the buttons are rendered
  it('should render all the buttons', () => {
    const { getByLabelText } = customRender(<Controls {...commonProps} />, {
      providerProps: { ...providerProps, enableDarkToggle: true },
    });
    expect(getByLabelText('first')).toBeInTheDocument();
    expect(getByLabelText('previous')).toBeInTheDocument();
    expect(getByLabelText('next')).toBeInTheDocument();
    expect(getByLabelText('last')).toBeInTheDocument();
    expect(getByLabelText('dark')).toBeInTheDocument();
  });

  // should render the play button when slideShowEnabled is true
  it('should render the play button when slideShowEnabled is true', () => {
    const { getByLabelText } = customRender(
      <Controls {...commonProps} slideShowEnabled slideShowRunning={false} />,
      { providerProps },
    );
    expect(getByLabelText('start slideshow')).toBeInTheDocument();
  });

  // check if all the callbacks are executed as expected
  it('should execute the callbacks as expected', async () => {
    const onFirst = vi.fn();
    const onPrevious = vi.fn();
    const onNext = vi.fn();
    const onLast = vi.fn();
    const onToggleDarkMode = vi.fn();
    const user = userEvent.setup();

    const { getByLabelText } = customRender(
      <Controls
        {...commonProps}
        onFirst={onFirst}
        onPrevious={onPrevious}
        onNext={onNext}
        onLast={onLast}
        onToggleDarkMode={onToggleDarkMode}
      />,
      { providerProps: { ...providerProps, enableDarkToggle: true } },
    );
    await user.click(getByLabelText('first'));
    expect(onFirst).toHaveBeenCalled();
    await user.click(getByLabelText('previous'));
    expect(onPrevious).toHaveBeenCalled();
    await user.click(getByLabelText('next'));
    expect(onNext).toHaveBeenCalled();
    await user.click(getByLabelText('last'));
    expect(onLast).toHaveBeenCalled();
    await user.click(getByLabelText('dark'));
    expect(onToggleDarkMode).toHaveBeenCalled();
  });

  // check if the navigation controls to be hidden / not shown when disableInteraction is set to true
  it('should hide the navigation controls when disableInteraction is set to true', () => {
    const { queryByLabelText } = customRender(<Controls {...commonProps} />, {
      providerProps: { ...providerProps, disableInteraction: true },
    });
    expect(queryByLabelText('first')).not.toBeInTheDocument();
    expect(queryByLabelText('previous')).not.toBeInTheDocument();
    expect(queryByLabelText('next')).not.toBeInTheDocument();
    expect(queryByLabelText('last')).not.toBeInTheDocument();
  });
});
