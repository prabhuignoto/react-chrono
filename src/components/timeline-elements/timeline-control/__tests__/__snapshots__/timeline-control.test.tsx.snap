// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimelineControl > should render 1`] = `
<div>
  <div
    class="sc-hLQSwg RQUyL"
  >
    <ul
      class="sc-beySPh jDWHGu timeline-controls controls"
    >
      <li
        class="sc-guDLey iDgMta"
      >
        <button
          aria-controls="timeline-main-wrapper"
          aria-disabled="false"
          aria-label="first"
          class="sc-dmyCSP gJvEmF"
          data-test-id="jump-to-first"
          mode="VERTICAL_ALTERNATING"
          rotate="TRUE"
          tabindex="0"
          title="first"
        >
          <svg
            class="feather feather-chevrons-left"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="11 17 6 12 11 7"
            />
            <polyline
              points="18 17 13 12 18 7"
            />
          </svg>
        </button>
      </li>
      <li
        class="sc-guDLey iDgMta"
      >
        <button
          aria-controls="timeline-main-wrapper"
          aria-disabled="false"
          aria-label="previous"
          class="sc-dmyCSP gJvEmF"
          data-test-id="previous"
          mode="VERTICAL_ALTERNATING"
          rotate="TRUE"
          tabindex="0"
          title="previous"
        >
          <svg
            class="feather feather-chevron-left"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="15 18 9 12 15 6"
            />
          </svg>
        </button>
      </li>
      <li
        class="sc-guDLey iDgMta"
      >
        <button
          aria-controls="timeline-main-wrapper"
          aria-disabled="false"
          aria-label="next"
          class="sc-dmyCSP gJvEmF"
          data-test-id="next"
          mode="VERTICAL_ALTERNATING"
          rotate="TRUE"
          tabindex="0"
          title="next"
        >
          <svg
            class="feather feather-chevron-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </button>
      </li>
      <li
        class="sc-guDLey iDgMta"
      >
        <button
          aria-controls="timeline-main-wrapper"
          aria-disabled="false"
          aria-label="last"
          class="sc-dmyCSP gJvEmF"
          data-test-id="jump-to-last"
          mode="VERTICAL_ALTERNATING"
          rotate="TRUE"
          tabindex="0"
          title="last"
        >
          <svg
            class="feather feather-chevrons-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="13 17 18 12 13 7"
            />
            <polyline
              points="6 17 11 12 6 7"
            />
          </svg>
        </button>
      </li>
      <li
        class="sc-guDLey iDgMta"
      />
      <li
        class="sc-guDLey iDgMta"
      >
        <button
          aria-controls="timeline-main-wrapper"
          aria-label="dark"
          class="sc-dmyCSP goMsNJ"
          data-test-id="dark-toggle"
          tabindex="0"
          title="dark"
        >
          <svg
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
            />
          </svg>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
